### PR TITLE
Workspace features

### PR DESCRIPTION
**Pull request by kornelijepetak**
Changes related to #214, #215 as a work-in-progress.

Certain things are still to be noted:

* It's not determined how to persist `WorkspaceFeature` in the database (it has an undefined primary key `Id`).
* Sometimes totally unrelated test fail randomly if run in parallel (rate limiting?)
* This implementation goes along with a 3rd version discussed on #214, which means that if decision is to be made the other way, this PR will need to change accordingly.

TL;DR
I put this here not to lose the work I've done (during discussions) if it can be helpful later on.

### CODE REVIEW

**Comment by simonrozsival** [comment]
This PR, which closes #214 and also closes #215, is now ready for reviews.

Deserialization of the workspace features is implemented, as well as the API client for getting the list of workspaces features. Both integration and unit tests are included.

One notable thing is the interface `Toggl.Multivac.Models.IWorkspaceFeature`, which does not implement `IBaseModel` as the features do not have `Id`s. I also moved it from the `Models` directory into `SpecificModels` directory (better name suggestions welcome), so it is not used by generators (which look for the contents of the directory `Toggl.Multivac/Models`).

**Comment by simonrozsival** [comment]
I moved the `IWorkspaceFeature` to the common models' directory as @willsb requested.

File: Toggl.Ultrawave/ApiClients/WorkspaceFeaturesApi.cs [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+        internal class WorkspaceFeatureDTO
+        {
+            public WorkspaceFeatureId FeatureId { get; set; }
+            public string Name { get; set; }
+            public bool Enabled { get; set; }
+        }
+    }
+
+    internal static class WorkspaceFeatureCollectionExtensions
```
Comment (amulware) [inline Toggl.Ultrawave/ApiClients/WorkspaceFeaturesApi.cs]: Why extensions instead of methods on the objects above?

File: Toggl.Ultrawave/ApiClients/WorkspaceFeaturesApi.cs [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+        public IObservable<List<IWorkspaceFeature>> GetEnabledFeatures()
+        {
+            return CreateObservable<List<WorkspaceFeatureCollectionDTO>>(endPoints.Get, AuthHeader)
+                .Select(list =>
+                    list.ToWorkspaceFeatures()
+                        .Where(wf => wf.Enabled)
+                        .ToList());
+        }
+
+        public IObservable<List<IWorkspaceFeature>> GetEnabledFeaturesForWorkspace(long workspaceId)
```
Comment (amulware) [inline Toggl.Ultrawave/ApiClients/WorkspaceFeaturesApi.cs]: Why does this get all and then filter? There is an endpoint for getting the features of a specific workspace.

Also, what is this needed for? I don't see any use cases for our app.

File: Toggl.Ultrawave/ApiClients/WorkspaceFeaturesApi.cs [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+        public IObservable<List<IWorkspaceFeature>> GetAll()
+        {
+            return CreateObservable<List<WorkspaceFeatureCollectionDTO>>(endPoints.Get, AuthHeader)
+                .Select(list => list
+                    .ToWorkspaceFeatures()
+                    .ToList());
+        }
+
+        public IObservable<List<IWorkspaceFeature>> GetEnabledFeatures()
```
Comment (amulware) [inline Toggl.Ultrawave/ApiClients/WorkspaceFeaturesApi.cs]: I don't understand why this exists. If the consumer wants to filter, they can do that themselves.

However, I see no use case for the filtered list either.

File: Toggl.Ultrawave/ApiClients/WorkspaceFeaturesApi.cs [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+        {
+            public WorkspaceFeatureId FeatureId { get; set; }
+            public string Name { get; set; }
+            public bool Enabled { get; set; }
+        }
+    }
+
+    internal static class WorkspaceFeatureCollectionExtensions
+    {
+        internal static IEnumerable<IWorkspaceFeature> ToWorkspaceFeatures(this IEnumerable<WorkspaceFeatureCollectionDTO> collection)
```
Comment (amulware) [inline Toggl.Ultrawave/ApiClients/WorkspaceFeaturesApi.cs]: This returns a list, why not make the return type specific and avoid the wasteful `ToList()` above?

File: Toggl.Ultrawave.Tests.Integration/WorkspaceFeaturesEnumTests.cs [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+{
+    public class WorkspaceFeaturesIdEnumTests
+    {
+        public class BackendEnumValuesCheckTest : AuthenticatedEndpointBaseTests<List<(WorkspaceFeatureId FeatureId, string Name)>>
+        {
+            protected override IObservable<List<(WorkspaceFeatureId FeatureId, string Name)>> CallEndpointWith(ITogglApi togglApi)
+                => (togglApi.WorkspaceFeatures as WorkspaceFeaturesApi).GetAllRaw();
+
+            [Fact, LogTestInfo]
+            public async Task TestsWhetherEnumValuesMatchBackendResponse()
```
Comment (amulware) [inline Toggl.Ultrawave.Tests.Integration/WorkspaceFeaturesEnumTests.cs]: This test's name has different semantics than all our other tests.

File: Toggl.Ultrawave/ApiClients/WorkspaceFeaturesApi.cs (hunk 3 above)
Comment (simonrozsival) [inline Toggl.Ultrawave/ApiClients/WorkspaceFeaturesApi.cs]: This is original @kornelijepetak's implementation. I wasn't sure if you discussed it or not and I just couldn't find it. I will 🔥 it.

File: Toggl.Ultrawave/ApiClients/WorkspaceFeaturesApi.cs (hunk 2 above)
Comment (simonrozsival) [inline Toggl.Ultrawave/ApiClients/WorkspaceFeaturesApi.cs]: Same as above -> 🔥 .

File: Toggl.Ultrawave/ApiClients/WorkspaceFeaturesApi.cs (hunk 1 above)
Comment (simonrozsival) [inline Toggl.Ultrawave/ApiClients/WorkspaceFeaturesApi.cs]: The point is to have DTO objects just store the data without any business logic inside. The extension method is a bit of a hack, but it makes clear what the purpose of those objects is. I won't mind placing them inside the classes.

Comment (amulware) [inline Toggl.Ultrawave/ApiClients/WorkspaceFeaturesApi.cs]: I believe that would allow the classes to be private, right? That seems like a nicer solution to me, yeah.

Comment (simonrozsival) [inline Toggl.Ultrawave/ApiClients/WorkspaceFeaturesApi.cs]: It must be internal, because I need to access the DTO in tests: https://github.com/toggl/mobileapp/pull/295/files#diff-667bbee88f9bebceb2e74338711eeb93R26

I could expose the deserialization to the tests in some different way, but that means extra code just because of the tests, which I don't think is a better solution. Or I could remove the serialization test and say that deserialization is tested in fact in the integration tests... but that also doesn't feel right because this is the case of more complicated deserialization and the test in fact shows exactly why the DTO is needed and how is it used.

Comment (amulware) [inline Toggl.Ultrawave/ApiClients/WorkspaceFeaturesApi.cs]: Hmm. I see.

Comment (simonrozsival) [inline Toggl.Ultrawave/ApiClients/WorkspaceFeaturesApi.cs]: Does this mean you approve the extension method or do still request changes here?

Comment (amulware) [inline Toggl.Ultrawave/ApiClients/WorkspaceFeaturesApi.cs]: Oh, please move the methods. Otherwise, ok for now. I'm not super happy with it, but also am not sure how I would prefer it.

**Comment by simonrozsival** [comment]
I have rewritten the PR but as it changed significantly, I decided to close this one and open #386 instead.
